stan/math/opencl/kernels/tri_multiply.hpp

```diff
@@ -0,0 +1,155 @@
+
+      // indexes that determine the last indexes that need to compute
+      // in order to remove the unnecesary multiplications in the special
+      // multiplication of A*A^T
+      const int jMin = THREAD_BLOCK_SIZE * get_group_id(1);
+      const int iMax = THREAD_BLOCK_SIZE * get_group_id(0) + get_local_size(0);
+      // local memory
+      __local double A_local[THREAD_BLOCK_SIZE][THREAD_BLOCK_SIZE];
+      __local double B_local[THREAD_BLOCK_SIZE][THREAD_BLOCK_SIZE];
+
+      double acc[WORK_PER_THREAD];
+      for (int w = 0; w < WORK_PER_THREAD; w++) {
+        acc[w] = 0.0;
+      }
+
+      const int num_tiles = (K + THREAD_BLOCK_SIZE - 1) / THREAD_BLOCK_SIZE;
+
+      const int is_A_lower = lower_upper_A == LOWER;
+      const int is_A_upper = lower_upper_A == UPPER;
+      const int is_B_lower = lower_upper_B == LOWER;
+      const int is_B_upper = lower_upper_B == UPPER;
+      // This kernel is based on the kernel for the regular
+      // matrix multiply where the multiplitcation is split
+      // in tiles.
+      // In cases where a matrix is lower triangular
+      // its not necessary to multiply the elements
+      // over the diagonal, therefore those tiles
+      // in the matrix multiply can be skipped.
+      // With upper triangular matrices we dont need
+      // to multiply the elements under the diagonal,
+      // so thosee tiles can be skipped.
+      // The following code determines the start and
+      // end tile based on triangularity of the input matrices
+      // If no matrices are triangular the starting tile
+      // is 0 and the end tile is num_tiles-1 which
+      // is the same as in \link kernels/multiply.hpp multiply()
+      const int end_tile_A = !is_A_lower * (num_tiles - 1)
+                             + is_A_lower * (i / THREAD_BLOCK_SIZE);
+      const int end_tile_B = !is_B_upper * (num_tiles - 1)
+                             + is_B_upper * (j / THREAD_BLOCK_SIZE);
+      const int start_tile_A = is_A_upper * (i / THREAD_BLOCK_SIZE);
+      const int start_tile_B = is_B_lower * (j / THREAD_BLOCK_SIZE);
+      int start_tile = 0;
+      int end_tile = num_tiles - 1;
+      if (start_tile_A > start_tile_B) {
+        start_tile = start_tile_A;
+      } else {
+        start_tile = start_tile_B;
+      }
+      if (end_tile_A > end_tile_B) {
+        end_tile = end_tile_B;
+      } else {
+        end_tile = end_tile_A;
+      }
+      for (int tile_ind = start_tile; tile_ind <= end_tile; tile_ind++) {
+        const int tiled_i = THREAD_BLOCK_SIZE * tile_ind + thread_block_row;
+        const int tiled_j = THREAD_BLOCK_SIZE * tile_ind + thread_block_col;
+        // each thread copies WORK_PER_THREAD values to the local
+        // memory
+        for (int w = 0; w < WORK_PER_THREAD; w++) {
+          const int tiled_i = THREAD_BLOCK_SIZE * tile_ind + thread_block_row;
+          const int tiled_j = THREAD_BLOCK_SIZE * tile_ind + thread_block_col;
```

Comment:
should these be here? they are declared above

Reply:
my bad :/